File: Largest Color Value in a Directed Graph/Largest Color Value in a Directed Graph.java
// Solution for Largest Color Value in a Directed Graph in JAVA

class Solution {
    public int largestPathValue(String colors, int[][] edges) {
        int n = colors.length();
        Map<Integer, List<Integer>> adj = new HashMap<>();
        int[] indegree = new int[n];
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            adj.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
            indegree[v]++;
        }

        int[][] colorFreq = new int[n][26];
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
                colorFreq[i][colors.charAt(i) - 'a'] = 1;
            }
        }

        int maxColorCount = 0, visitedNodes = 0;
        while (!queue.isEmpty()) {
            int curr = queue.poll();
            visitedNodes++;
            maxColorCount = Math.max(maxColorCount, colorFreq[curr][colors.charAt(curr) - 'a']);
            for (int neighbor : adj.getOrDefault(curr, new ArrayList<>())) {
                for (int c = 0; c < 26; c++) {
                    int add = (colors.charAt(neighbor) - 'a' == c) ? 1 : 0;
                    colorFreq[neighbor][c] = Math.max(colorFreq[neighbor][c], colorFreq[curr][c] + add);
                }
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) queue.offer(neighbor);
            }
        }

        return visitedNodes == n ? maxColorCount : -1;
    }
}

File: Largest Color Value in a Directed Graph/Largest Color Value in a Directed Graph.py
# Solution for Largest Color Value in a Directed Graph in PY

class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        n = len(colors)
        adj = defaultdict(list)
        indegree = [0] * n
        for u, v in edges:
            adj[u].append(v)
            indegree[v] += 1

        color_freq = [[0] * 26 for _ in range(n)]
        queue = deque()
        for i in range(n):
            if indegree[i] == 0:
                queue.append(i)
                color_freq[i][ord(colors[i]) - ord('a')] = 1

        max_color_count = 0
        visited_nodes = 0
        while queue:
            curr = queue.popleft()
            visited_nodes += 1
            max_color_count = max(max_color_count, color_freq[curr][ord(colors[curr]) - ord('a')])
            for neighbor in adj[curr]:
                for c in range(26):
                    add = 1 if ord(colors[neighbor]) - ord('a') == c else 0
                    color_freq[neighbor][c] = max(color_freq[neighbor][c], color_freq[curr][c] + add)
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)

        return max_color_count if visited_nodes == n else -1
